### PR TITLE
document the plugin api for pre_configure and post_install

### PR DESCRIPTION
**Pull request by wxtim**
Replaces https://github.com/cylc/cylc-flow/pull/4962

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.

### CODE REVIEW

File: src/plugins/install/index.rst [hunk 1]
```diff
@@ -1,4 +1,157 @@
...
+---------------------------------------------------------
+
+Cylc uses entry points registered by setuptools to search for pre-configure
+and post-install plugins.
+
+Hello World
+^^^^^^^^^^^
+
+In this example a pre-configure plugin which logs a "Hello World" message,
+and, after installation logs some info about the installation:
```
Comment (MetRonnie) [inline src/plugins/install/index.rst:35]: (The comma nazi has logged on)
```suggestion
In this example a pre-configure plugin which logs a "Hello World" message
and, after installation, logs some info about the installation:
```

File: src/plugins/install/index.rst [hunk 2]
```diff
@@ -1,4 +1,157 @@
...
+
+.. code-block:: python
+   :caption: my_plugin.py
+
+   from cylc.flow import LOG
+
+   def pre_configure(srcdir=None, options=None, rundir=None):
```
Comment (MetRonnie) [inline src/plugins/install/index.rst:42]:
```suggestion
   def pre_configure(srcdir=None, opts=None, rundir=None):
```

File: src/plugins/install/index.rst [hunk 3]
```diff
@@ -1,4 +1,157 @@
...
+            # name = python.namespace.of.module
+            'my_plugin=my_plugin.my_plugin:post_install'
+          ]
+       }
+   )
+
+API Reference
+^^^^^^^^^^^^^
+
+Cylc will pass following variables to pre-configure and post-install plugins:
```
Comment (MetRonnie) [inline src/plugins/install/index.rst:84]:
```suggestion
Cylc will pass following arguments to pre-configure and post-install plugins:
```

File: src/plugins/install/index.rst [hunk 4]
```diff
@@ -1,4 +1,157 @@
...
+
+``srcdir`` (Path or string)
+   The directory from which ``cylc install`` is installing a workflow,
+   or the directory passed to ``cylc validate``, ``cylc graph`` and other
+   CLI commands which work without installing a workflow.
+``options`` (Namespace)
```
Comment (MetRonnie) [inline src/plugins/install/index.rst:90]:
```suggestion
``opts`` (``optparse.Values``)
```

File: src/plugins/install/index.rst [hunk 5]
```diff
@@ -1,4 +1,157 @@
...
+``rundir`` (Path or string)
+   The destination of a ``cylc install`` or ``reinstall`` command.
+
+The pre-confugure plugin should return a dictionary which may contain the
+following keys:
+
+``env``
+   A dictionary of environment variables to be exported to the scheduler
+   environment.
+``template variables``
```
Comment (MetRonnie) [inline src/plugins/install/index.rst:101]:
```suggestion
``template_variables``
```

**Comment by MetRonnie** [comment]
This has replaced `cylc/flow/install_plugins/__init__.py`, so that can be deleted

File: src/plugins/install/index.rst [hunk 6]
```diff
@@ -1,4 +1,157 @@
...
+   A dictionary of template variables to be used by Jinja2 or Empy when
+   templating the workflow configuration files.
+``templating_detected``
+   ``jinja2`` | ``empy`` to be used when templating. N.b: This will result in
+   failure if the templating language set does not match the shebang line of
+   the ``flow.cylc`` file.
+
+The post-install entry point does not return any data used by Cylc.
+
+More advanced example
```
Comment (oliver-sanders) [inline src/plugins/install/index.rst:111]: Perhaps say, "see cylc-rose" for a concrete implementation.

**Review by oliver-sanders — APPROVED** [review]

**Comment by oliver-sanders** [comment]
Note:
* The original issue is not linked to this PR.
* @MetRonnie made a suggestion on https://github.com/cylc/cylc-flow/pull/4962#discussion_r925399582

**Comment by wxtim** [comment]
> This has replaced `cylc/flow/install_plugins/__init__.py`, so that can be deleted

I'm not sure that's accurate -  This document doesn't talk about the single built in plugin listed at https://cylc.github.io/cylc-doc/nightly/html/plugins/install/index.html

File: src/plugins/install/index.rst (hunk 6 above)
Comment (wxtim) [reply src/plugins/install/index.rst:111]: Instead of this section, or as well as?

File: src/plugins/install/index.rst [hunk 7]
```diff
@@ -1,4 +1,157 @@
...
+^^^^^^^^^^^^^^^^^^^^^
+
```
Comment (wxtim) [inline src/plugins/install/index.rst:113]:
```suggestion

.. seealso::

   For the implementation of a more fully featured "real-world" example see
   :ref:`Cylc-Rose`.

```

**Comment by MetRonnie** [comment]
> /home/runner/work/cylc-doc/cylc-doc/src/plugins/install/index.rst:116:undefined label: cylc-rose

**Comment by MetRonnie** [comment]
> > This has replaced `cylc/flow/install_plugins/__init__.py`, so that can be deleted
> 
> I'm not sure that's accurate - This document doesn't talk about the single built in plugin listed at https://cylc.github.io/cylc-doc/nightly/html/plugins/install/index.html

Yes it does...
https://github.com/cylc/cylc-doc/blob/a18694f809ccf49a8fd07ce17cbbff9ac68eed59/src/plugins/install/index.rst#L11-L23

File: src/plugins/install/index.rst [hunk 8]
```diff
@@ -1,4 +1,162 @@
...
+``srcdir`` (Path or string)
+   The directory from which ``cylc install`` is installing a workflow,
+   or the directory passed to ``cylc validate``, ``cylc graph`` and other
+   CLI commands which work without installing a workflow.
+``opts`` (``optparse.Values``)
+   CLI options set for a Cylc script.
+``rundir`` (Path or string)
+   The destination of a ``cylc install`` or ``reinstall`` command.
+
+The pre-confugure plugin should return a dictionary which may contain the
```
Comment (MetRonnie) [inline src/plugins/install/index.rst:95]:
```suggestion
The pre-configure plugin should return a dictionary which may contain the
```

File: src/plugins/install/index.rst [hunk 9]
```diff
@@ -1,4 +1,162 @@
...
+          'cylc.pre_configure': [
+            # name = python.namespace.of.module
+            'my_plugin=my_plugin.my_plugin:pre_configure'
+          ]
+          'cylc.post_install': [
+            # name = python.namespace.of.module
+            'my_plugin=my_plugin.my_plugin:post_install'
+          ]
+       }
+   )
```
Comment (MetRonnie) [inline src/plugins/install/index.rst:79]: Python indentation not quite right for these examples

File: src/plugins/install/index.rst [hunk 10]
```diff
@@ -1,4 +1,162 @@
...
+   The destination of a ``cylc install`` or ``reinstall`` command.
+
+The pre-confugure plugin should return a dictionary which may contain the
+following keys:
+
+``env``
+   A dictionary of environment variables to be exported to the scheduler
+   environment.
+``template_variables``
+   A dictionary of template variables to be used by Jinja2 or Empy when
```
Comment (MetRonnie) [inline src/plugins/install/index.rst:102]:
```suggestion
   A dictionary of template variables to be used by Jinja2 or EmPy when
```

File: src/plugins/install/index.rst [hunk 11]
```diff
@@ -1,4 +1,162 @@
...
+               'templating_detected': 'jinja2'
+         }
+      else:
+         return {}
+
+   def post_install(srcdir=None, opts=None, rundir=None):
+      # record plugin version in a file
+      (Path(rundir) / 'log/json-plugin.info').write_text(
+         f"Installed with Simple JSON reader plugin version {VERSION}\\n")
+      return None
```
Comment (MetRonnie) [inline src/plugins/install/index.rst:162]: Indentation

File: src/plugins/install/index.rst (hunk 9 above)
Comment (wxtim) [reply src/plugins/install/index.rst]: Gah!

**Comment by wxtim** [comment]
Thanks @MetRonnie - I think that I've fixed the python indentation issues.

**Comment by MetRonnie** [comment]
Still failing with
> > /home/runner/work/cylc-doc/cylc-doc/src/plugins/install/index.rst:116:undefined label: cylc-rose

**Review by MetRonnie — APPROVED** [review]
Apart from the fact that it looks like there are two identical commit histories merged into the same branch, I think this is good to go

**Comment by wxtim** [comment]
> Apart from the fact that it looks like there are two identical commit histories merged into the same branch, I think this is good to go

Looks horribly polluted to me...

**Review by MetRonnie — APPROVED** [review]

**Comment by wxtim** [comment]
poking @oliver-sanders  for a re-review

**Review by oliver-sanders — APPROVED** [review]

**Pull request merged**